ethereumj-core/src/main/java/org/ethereum/listener/EventListener.java

```diff
@@ -0,0 +1,304 @@
+        }
+    }
+
+    protected LogFilter contractAddressFilter;
+    protected CallTransaction.Contract contract;
+    protected PendingStateImpl pendingState;
+
+    // txHash => PendingEvent
+    protected Map<ByteArrayWrapper, PendingEvent> pendings = new LinkedHashMap<>();
+    protected Block bestBlock;
+    BigInteger lastTotDiff = BigInteger.ZERO;
+    // executing EthereumListener callbacks on a separate thread to avoid long running
+    // handlers (die to DB access) messing up core
+    ExecutorService executor = Executors.newSingleThreadExecutor(r -> new Thread(r, EventListener.this.getClass().getSimpleName() + "-exec"));
+
+    public final EthereumListener listener =  new EthereumListenerAdapter() {
+        @Override
+        public void onBlock(BlockSummary blockSummary) {
+            executor.submit(() -> onBlockImpl(blockSummary));
+        }
+
+        @Override
+        public void onPendingTransactionUpdate(TransactionReceipt txReceipt, PendingTransactionState state, Block block) {
+            executor.submit(() -> onPendingTransactionUpdateImpl(txReceipt, state, block));
+        }
+    };
+
+    public EventListener(PendingStateImpl pendingState) {
+        this.pendingState = pendingState;
+    }
+
+    public void onBlockImpl(BlockSummary blockSummary) {
+        try {
+            logger.debug("onBlock: " + blockSummary.getBlock().getShortDescr());
+
+            // ignoring spurious old blocks
+            if (bestBlock != null && blockSummary.getBlock().getNumber() < bestBlock.getNumber() - 16) {  // FIXME: why 16?
+                logger.debug("Ignoring block as too old: " + blockSummary.getBlock().getShortDescr());
+                return;
+            }
+
+            if (contractAddressFilter.matchBloom(new Bloom(blockSummary.getBlock().getLogBloom()))) {
+                for (int i = 0; i < blockSummary.getReceipts().size(); i++) {
+                    TransactionReceipt receipt = blockSummary.getReceipts().get(i);
+                    if (contractAddressFilter.matchBloom(receipt.getBloomFilter())) {
+                        if (!pendings.containsKey(new ByteArrayWrapper(receipt.getTransaction().getHash()))) {
+                            // ask PendingState to track candidate transactions from blocks since there is
+                            // no guarantee the transaction of interest received as a pending
+                            // on included transaction PendingState should call onPendingTransactionUpdateImpl
+                            // with INCLUDED state
+                            // if Tx was included into fork block PendingState should call onPendingTransactionUpdateImpl
+                            // with PENDING state
+                            pendingState.trackTransaction(receipt.getTransaction());
+                        }
+                    }
+                }
+            }
+
+            if (blockSummary.getTotalDifficulty().compareTo(lastTotDiff) > 0) {
```

Comment:
Isn't it a better idea to supply `BlockSummary` with `ImportStatus` to indicate whether imported block is best or not. This small feature might be useful in many other places and save us from keeping this redundant code and data structures

Reply:
Good idea, but I'd rather put this logic in BlockSummary, because ImportState couldn't be always kept in BlockSummary, as summary may be null. I'll push this approach and we could discuss it, if it doesn't look ok.